File: packages/shared/src/types/billing.ts
/** Invoice status lifecycle. */
export type InvoiceStatus =
  | "draft"
  | "finalized"
  | "partially_paid"
  | "paid"
  | "cancelled"
  | "waived";

/** Payment method options. */
export type PaymentMethod = "cash" | "mpesa" | "insurance" | "exemption";

/** Invoice line item type. */
export type InvoiceItemType =
  | "consultation"
  | "pharmacy"
  | "lab"
  | "procedure"
  | "bed"
  | "nursing"
  | "other";

// ── Invoice ──────────────────────────────────────────────────────────────────

/** Invoice response from the API. */
export interface InvoiceResponse {
  id: string;
  encounter_id: string;
  patient_id: string;
  invoice_number: string;
  status: InvoiceStatus;
  payment_method: PaymentMethod | null;
  insurance_provider: string | null;
  insurance_member_no: string | null;
  sha_claim_ref: string | null;
  subtotal_cents: number;
  discount_cents: number;
  tax_cents: number;
  total_cents: number;
  paid_cents: number;
  balance_cents: number;
  notes: string | null;
  finalized_at: string | null;
  created_at: string;
}

/** Invoice line item response. */
export interface InvoiceItemResponse {
  id: string;
  invoice_id: string;
  item_type: InvoiceItemType;
  description: string;
  quantity: number;
  unit_price_cents: number;
  total_cents: number;
  discount_cents: number;
  reference_id: string | null;
  reference_type: string | null;
}

/** Full invoice with items and patient info. */
export interface InvoiceDetail {
  invoice: InvoiceResponse;
  items: InvoiceItemResponse[];
  patient_name: string | null;
  patient_mrn: string | null;
}

/** Invoice list item (summary). */
export interface InvoiceListItem {
  id: string;
  invoice_number: string;
  encounter_id: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  status: InvoiceStatus;
  total_cents: number;
  paid_cents: number;
  balance_cents: number;
  payment_method: PaymentMethod | null;
  item_count: number;
  created_at: string;
}

/** Paginated invoice list. */
export interface InvoiceListResponse {
  items: InvoiceListItem[];
  total: number;
  page: number;
  page_size: number;
}

// ── Invoice Create ───────────────────────────────────────────────────────────

/** Payload for creating a line item. */
export interface InvoiceItemCreate {
  item_type: InvoiceItemType;
  description: string;
  quantity: number;
  unit_price_cents: number;
  discount_cents?: number;
  reference_id?: string | null;
  reference_type?: string | null;
}

/** Payload for creating an invoice. */
export interface InvoiceCreate {
  encounter_id: string;
  patient_id: string;
  payment_method?: PaymentMethod | null;
  insurance_provider?: string | null;
  insurance_member_no?: string | null;
  notes?: string | null;
  items: InvoiceItemCreate[];
}

// ── Payment ──────────────────────────────────────────────────────────────────

/** Payload for recording a payment. */
export interface PaymentCreate {
  amount_cents: number;
  payment_method: PaymentMethod;
  reference_number?: string | null;
  mpesa_transaction_id?: string | null;
  notes?: string | null;
}

/** Payment response from the API. */
export interface PaymentResponseType {
  id: string;
  invoice_id: string;
  patient_id: string;
  amount_cents: number;
  payment_method: PaymentMethod;
  reference_number: string | null;
  mpesa_transaction_id: string | null;
  received_by: string;
  paid_at: string;
  notes: string | null;
}

// ── Dashboard ────────────────────────────────────────────────────────────────

/** Billing dashboard summary. */
export interface BillingSummary {
  total_invoices: number;
  draft_count: number;
  finalized_count: number;
  paid_count: number;
  partially_paid_count: number;
  total_billed_cents: number;
  total_paid_cents: number;
  total_outstanding_cents: number;
}

/** Invoice waiver request. */
export interface InvoiceWaiveRequest {
  reason: string;
}

File: services/api-gateway/app/schemas/billing.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

# ── Invoice schemas ───────────────────────────────────────────────────────────


class InvoiceItemCreate(BaseModel):
    """Schema for adding a line item to an invoice."""

    item_type: str = Field(
        ...,
        pattern=r"^(consultation|pharmacy|lab|procedure|bed|nursing|other)$",
    )
    description: str = Field(..., min_length=1, max_length=500)
    quantity: int = Field(default=1, ge=1)
    unit_price_cents: int = Field(default=0, ge=0)
    discount_cents: int = Field(default=0, ge=0)
    reference_id: uuid.UUID | None = None
    reference_type: str | None = Field(None, pattern=r"^(dispensing|lab_order|encounter)$")


class InvoiceCreate(BaseModel):
    """Schema for creating an invoice."""

    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    payment_method: str | None = Field(None, pattern=r"^(cash|mpesa|insurance|exemption)$")
    insurance_provider: str | None = Field(None, max_length=100)
    insurance_member_no: str | None = Field(None, max_length=50)
    notes: str | None = Field(None, max_length=2000)
    items: list[InvoiceItemCreate] = Field(..., min_length=1)


class InvoiceItemResponse(BaseModel):
    """Schema for invoice item API responses."""

    id: uuid.UUID
    invoice_id: uuid.UUID
    item_type: str
    description: str
    quantity: int
    unit_price_cents: int
    total_cents: int
    discount_cents: int
    reference_id: uuid.UUID | None
    reference_type: str | None

    model_config = {"from_attributes": True}


class InvoiceResponse(BaseModel):
    """Schema for invoice API responses."""

    id: uuid.UUID
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    invoice_number: str
    status: str
    payment_method: str | None
    insurance_provider: str | None
    insurance_member_no: str | None
    sha_claim_ref: str | None
    subtotal_cents: int
    discount_cents: int
    tax_cents: int
    total_cents: int
    paid_cents: int
    balance_cents: int
    notes: str | None
    finalized_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class InvoiceDetail(BaseModel):
    """Full invoice with line items and patient info."""

    invoice: InvoiceResponse
    items: list[InvoiceItemResponse]
    patient_name: str | None = None
    patient_mrn: str | None = None


class InvoiceListItem(BaseModel):
    """Invoice summary for list views."""

    id: uuid.UUID
    invoice_number: str
    encounter_id: uuid.UUID
    patient_id: uuid.UUID
    patient_name: str | None = None
    patient_mrn: str | None = None
    status: str
    total_cents: int
    paid_cents: int
    balance_cents: int
    payment_method: str | None
    item_count: int = 0
    created_at: datetime

    model_config = {"from_attributes": True}


class InvoiceListResponse(BaseModel):
    """Paginated invoice list response."""

    items: list[InvoiceListItem]
    total: int
    page: int
    page_size: int


# ── Payment schemas ───────────────────────────────────────────────────────────


class PaymentCreate(BaseModel):
    """Schema for recording a payment against an invoice."""

    amount_cents: int = Field(..., gt=0)
    payment_method: str = Field(..., pattern=r"^(cash|mpesa|insurance|exemption)$")
    reference_number: str | None = Field(None, max_length=100)
    mpesa_transaction_id: str | None = Field(None, max_length=50)
    notes: str | None = Field(None, max_length=2000)


class PaymentResponse(BaseModel):
    """Schema for payment API responses."""

    id: uuid.UUID
    invoice_id: uuid.UUID
    patient_id: uuid.UUID
    amount_cents: int
    payment_method: str
    reference_number: str | None
    mpesa_transaction_id: str | None
    received_by: uuid.UUID
    paid_at: datetime
    notes: str | None

    model_config = {"from_attributes": True}


# ── Billing Dashboard ─────────────────────────────────────────────────────────


class BillingSummary(BaseModel):
    """Dashboard summary stats."""

    total_invoices: int
    draft_count: int
    finalized_count: int
    paid_count: int
    partially_paid_count: int
    total_billed_cents: int
    total_paid_cents: int
    total_outstanding_cents: int


# ── Waive / Cancel ────────────────────────────────────────────────────────────


class InvoiceWaiveRequest(BaseModel):
    """Schema for waiving an invoice balance."""

    reason: str = Field(..., min_length=5, max_length=500)
